### PR TITLE
feat(83547): Adding escaping of property name for GetPath.

### PR DESCRIPTION
**Pull request by Maximys**
Current PR fix problem with unescaped single quote, but it does not fix problem with shorthand property.
I did so because [Newtonsoft.Json 13.0.3](https://github.com/JamesNK/Newtonsoft.Json/releases/tag/13.0.3) have the same behavior:
```
JToken jToken;

jToken = JToken.Parse("""{"$myRoot": {"myRoot'child": {"secondLevelChildWithoutEscaping": "value1"}}}""");

jToken = jToken["$myRoot"]["myRoot'child"]["secondLevelChildWithoutEscaping"];

Console.WriteLine(jToken.Path);
```

output: _"$myRoot['myRoot\'child'].secondLevelChildWithoutEscaping"_

I think, Newtonsoft is is etalon implementation of JSON library, but nobody of us protected from mistake.
Add some comment if this behaviour invalid.

Contributes to #83547

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs [hunk 1]
```diff
@@ -53,6 +53,40 @@ public static byte[] GetEscapedPropertyNameSection(ReadOnlySpan<byte> utf8Value,
...
+
+        private static string GetEscapedPropertyName(string rawPropertyName)
+        {
+            const string SingleQuote = "'";
+            const string EscapedSingleQuote = "\\'";
+
+            string returnValue;
+
+            returnValue = rawPropertyName
+                .Replace(SingleQuote, EscapedSingleQuote);
```
Comment (Maximys) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:85]: I wanted to use here [Regex.Replace(String, String, String, RegexOptions)](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.replace?view=net-7.0#system-text-regularexpressions-regex-replace(system-string-system-string-system-string-system-text-regularexpressions-regexoptions)), but I could not understand how to use System.Text.RegularExpressions assembly from current.

**Comment by Maximys** [comment]
@layomia , could you review current PR?

**Comment by tarekgh** [comment]
@Maximys looks there are some related test failures because of this change. Could you please have a look?

**Comment by gregsdennis** [comment]
> Fixes #83547

If the shorthand problem isn't being addressed, the issue should remain open.  This PR isn't a complete fix.

**Comment by Maximys** [comment]
@tarekgh , ok, I'll try to look

**Comment by tarekgh** [comment]
@layomia @eiriktsarpalis could you please have a look at this change. The change is reasonable to me, but I want to ensure this is not breaking in anyway. Thanks!

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 2]
```diff
@@ -394,6 +394,20 @@ public Parser(Compilation compilation)
                 };
             }
 
+            private static string EscapeString(string representation)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:397]: This appears kind of ad-hoc. Consider using the built-in [SytaxFactory.Literal](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.visualbasic.syntaxfactory.literal?view=roslyn-dotnet-4.6.0#microsoft-codeanalysis-visualbasic-syntaxfactory-literal(system-string)) factory to escape the string as expected.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:397]: Please update the `FormatStringLiteral` method in the emitter class so that `SyntaxFactory` is used there:

https://github.com/dotnet/runtime/blob/45865877ecd3d1506e0aa19d810fb35bed4b5f9e/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs#L1239

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs [hunk 3]
```diff
@@ -53,6 +53,43 @@ public static byte[] GetEscapedPropertyNameSection(ReadOnlySpan<byte> utf8Value,
             return escapedString;
         }
 
+        public static string GetPropertyName(
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:56]: Unlike the source generator, escaping shouldn't apply to runtime components. This is already handled by `JsonSerializerOptions.Encoder` property which by default escapes characters to their unicode representation:

```C#
Console.WriteLine(JsonSerializer.Serialize(new Foo { Name = "test" })); // {"\u0022\u0027":"test"}

public class Foo
{
    [JsonPropertyName("\"'")]
    public string Name { get; set; }
}
```

Any fix to JsonNode.Path should try to be consistent with what the serializer is doing already, i.e. encode single-quote characters to \u0027.

**Review by eiriktsarpalis — CHANGES_REQUESTED** [review]
Escaping should consistent with what `JsonSerializerOptions.Encoder` is doing.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 2 above)
Comment (Maximys) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:397]: @eiriktsarpalis , I'll try to understand how to use [SytaxFactory.Literal](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.visualbasic.syntaxfactory.literal?view=roslyn-dotnet-4.6.0#microsoft-codeanalysis-visualbasic-syntaxfactory-literal(system-string)) on weekend and apply required changes.
You had provided link to Microsoft.CodeAnalysis.**VisualBasic**.SyntaxFactory.Literal. Are you sure, that it's correct? Why not Microsoft.CodeAnalysis.**CSharp**.SyntaxFactory.Literal?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:397]: Apologies, of course it should be the `CSharp` equivalent.

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs (hunk 3 above)
Comment (Maximys) [reply src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:56]: @eiriktsarpalis , fixed by https://github.com/dotnet/runtime/pull/84688/commits/338a084cdde4008fc4d7d32eb370cd0fd5351501

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 2 above)
Comment (Maximys) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs]: Fixed by https://github.com/dotnet/runtime/pull/84688/commits/75dedede15be085d7fb2cb3ab4a8da78acfd0514

**Comment by Maximys** [comment]
@eiriktsarpalis , could I ask little question, which is not linked with current PR? I want try to fix #86031 by separate PR. I want to use [FluentAssertions](https://www.nuget.org/packages/FluentAssertions) for check result of my fix. As I can [see](https://github.com/dotnet/runtime/blob/d638e1e012f2e15feb98eb1099fa33cec00e1768/src/installer/tests/TestUtils/TestUtils.csproj#L15), Rutime currently use this library only for test [installer](https://github.com/dotnet/runtime/tree/main/src/installer). May I add FluentAssertions to TestUtilities inside [Common](https://github.com/dotnet/runtime/tree/main/src/libraries/Common/tests/TestUtilities)?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 4]
```diff
@@ -1236,7 +1236,21 @@ private static string GetUnmappedMemberHandlingAsStr(JsonUnmappedMemberHandling
...
+                {
+                    returnValue = "null";
+                }
+                else
+                {
+                    returnValue = SyntaxFactory.Literal(value).ToFullString();
+                }
+
+                return returnValue;
+            }
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:1253]: Seems like this can be simplified
```suggestion
                => value is null ? "null" : SyntaxFactory.Literal(value).ToFullString();
```

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 5]
```diff
@@ -1000,7 +1000,6 @@ private bool IsValidDataExtensionPropertyType(ITypeSymbol type)
                                 {
                                     ImmutableArray<TypedConstant> ctorArgs = attributeData.ConstructorArguments;
                                     jsonPropertyName = (string)ctorArgs[0].Value!;
-                                    // Null check here is done at runtime within JsonSerializer.
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1003]: Is this change related?

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs [hunk 6]
```diff
@@ -53,6 +53,34 @@ public static byte[] GetEscapedPropertyNameSection(ReadOnlySpan<byte> utf8Value,
...
+            string rawPropertyName,
+            JavaScriptEncoder? encoder)
+        {
+            int indexOfFirstCharacterToEncode;
+            string returnValue;
+
+            indexOfFirstCharacterToEncode = JsonWriterHelper.NeedsEscaping(rawPropertyName.AsSpan(), encoder);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:63]: Given that this method is only invoked after a check to `propertyName.AsSpan().ContainsSpecialCharacters()` this check seems a bit redundant to me. It seems like this could be replaced with
```suggestion
            Debug.Assert(JsonWriterHelper.NeedsEscaping(rawPropertyName.AsSpan(), encoder) != -1);
```
And the method renamed to `GetEscapedPropertyName` (like the helper you have defined below).

**Comment by eiriktsarpalis** [comment]
Hi @Maximys, apologies for the delay in reviewing this. Any chance you could rebase your changes and address a few other small pieces of feedback? Thanks!

**Comment by Maximys** [comment]
@eiriktsarpalis , yes, I'll try to make it

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs (hunk 4 above)
Comment (Maximys) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs]: Fixed by https://github.com/dotnet/runtime/pull/84688/commits/51fcbb09a8b9c5eee71d151817455c97bf484545

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 5 above)
Comment (Maximys) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs]: No, fixed by amend

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs (hunk 6 above)
Comment (Maximys) [reply src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:63]: > Given that this method is only invoked after a check to `propertyName.AsSpan().ContainsSpecialCharacters()` this check seems a bit redundant to me. It seems like this could be replaced with
> 
> And the method renamed to `GetEscapedPropertyName` (like the helper you have defined below).

I fixed it by https://github.com/dotnet/runtime/pull/84688/commits/a610e38ee705bfe38420a1993af2ffc6b8101186 , but after that **GetPath_Special Characters** test was failed with next error message:

> [xUnit.net 00:06:26.33]     System.Text.Json.Nodes.Tests.ParentPathRootTests.GetPath_SpecialCharacters [FAIL]
>   Failed System.Text.Json.Nodes.Tests.ParentPathRootTests.GetPath_SpecialCharacters [626 ms]
>   Error Message:
>    Microsoft.VisualStudio.TestPlatform.TestHost.DebugAssertException : Method Debug.Fail failed with '
> ', and was translated to Microsoft.VisualStudio.TestPlatform.TestHost.DebugAssertException to avoid terminating the process hosting the test.
>   Stack Trace:
>      at Microsoft.VisualStudio.TestPlatform.TestHost.TestHostTraceListener.GetException(String message)
>    at Microsoft.VisualStudio.TestPlatform.TestHost.TestHostTraceListener.Fail(String message, String detailMessage)
>    at System.Diagnostics.TraceInternal.Fail(String message, String detailMessage) in C:\runtime\src\libraries\System.Diagnostics.TraceSource\src\System\Diagnostics\TraceInternal.cs:line 262
>    at System.Diagnostics.TraceInternal.TraceProvider.Fail(String message, String detailMessage) in C:\runtime\src\libraries\System.Diagnostics.TraceSource\src\System\Diagnostics\TraceInternal.cs:line 17
>    at System.Diagnostics.Debug.Fail(String message, String detailMessage) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Diagnostics\Debug.cs:line 133
>    at System.Diagnostics.Debug.Assert(Boolean condition, String message, String detailMessage) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Diagnostics\Debug.cs:line 97
>    at System.Diagnostics.Debug.Assert(Boolean condition) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Diagnostics\Debug.cs:line 82
>    at System.Text.Json.JsonHelpers.GetEscapedPropertyName(String rawPropertyName, JavaScriptEncoder encoder) in C:\runtime\src\libraries\System.Text.Json\src\System\Text\Json\JsonHelpers.Escaping.cs:line 60
>    at System.Text.Json.Nodes.JsonObject.GetPath(ValueStringBuilder& path, JsonNode child) in C:\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Nodes\JsonObject.cs:line 210
>    at System.Text.Json.Nodes.JsonValue.GetPath(ValueStringBuilder& path, JsonNode child) in C:\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Nodes\JsonValue.cs:line 97
>    at System.Text.Json.Nodes.JsonNode.GetPath() in C:\runtime\src\libraries\System.Text.Json\src\System\Text\Json\Nodes\JsonNode.cs:line 136
>    at System.Text.Json.Nodes.Tests.ParentPathRootTests.GetPath_SpecialCharacters() in C:\runtime\src\libraries\System.Text.Json\tests\System.Text.Json.Tests\JsonNode\ParentPathRootTests.cs:line 99
>    at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
>    at System.Reflection.MethodBaseInvoker.InterpretedInvoke_Method(Object obj, IntPtr* args) in C:\runtime\src\coreclr\System.Private.CoreLib\src\System\Reflection\MethodBaseInvoker.CoreCLR.cs:line 36
>    at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodBaseInvoker.cs:line 57
>    at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Reflection\RuntimeMethodInfo.cs:line 127
>    at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Reflection\MethodBase.cs:line 56
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38
>    at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38                                                                                                                                                                                                  at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38                                                                                                                                                                                                  at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38                                                                                                                                                                                          10:19   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38                                                                                                                                                                                                  at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38                                                                                                                                                                                                  at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38                                                                                                                                                                                                  at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\CompilerServices\AsyncMethodBuilderCore.cs:line 38                                                                                                                                                                                                  at System.Threading.Tasks.Task`1.InnerInvoke() in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Future.cs:line 501                                                                    at System.Threading.Tasks.Task.<>c.<.cctor>b__281_0(Object obj) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 2387                                                    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\ExecutionContext.cs:line 179                                                                                                                                                                                                      at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 2345              at System.Threading.Tasks.Task.ExecuteEntry() in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 2258                                                                      at System.Threading.Tasks.SynchronizationContextTaskScheduler.<>c.<.cctor>b__8_0(Object s) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\TaskScheduler.cs:line 606                 at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\ExecutionContext.cs:line 179                                                                                                                                                                                                      at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\ExecutionContext.cs:line 143                                                                                                                                                                                                              at System.Threading.Tasks.Task.InnerInvoke() in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 2405                                                                       at System.Threading.Tasks.Task.<>c.<.cctor>b__281_0(Object obj) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 2387                                                    at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\ExecutionContext.cs:line 179                                                                                                                                                                                                      at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 2345              at System.Threading.Tasks.Task.ExecuteEntryUnsafe(Thread threadPoolThread) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\Task.cs:line 2283                                         at System.Threading.Tasks.ThreadPoolTaskScheduler.<>c.<.cctor>b__10_0(Object s) in C:\runtime\src\libraries\System.Private.CoreLib\src\System\Threading\Tasks\ThreadPoolTaskScheduler.cs:line 35                   at System.Threading.Thread.StartCallback() in C:\runtime\src\coreclr\System.Private.CoreLib\src\System\Threading\Thread.CoreCLR.cs:line 104

That's why I made revert https://github.com/dotnet/runtime/pull/84688/commits/4a0d7b17c4ad86dbe77fce5281a41bbb175e53d9

**Comment by gregsdennis** [comment]
> > Fixes #83547
> 
> If the shorthand problem isn't being addressed, the issue should remain open.  This PR isn't a complete fix.

@Maximys of this isn't going to be addressed, can you remove the "fixes" text from the opening comment, please?

Alternatively, a secondary issue should be opened to cover the shorthand problem.

File: src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs [hunk 7]
```diff
@@ -53,6 +53,34 @@ public static byte[] GetEscapedPropertyNameSection(ReadOnlySpan<byte> utf8Value,
...
+            return returnValue;
+        }
+
+        private static string GetEscapedPropertyName(
+            string rawPropertyName,
+            JavaScriptEncoder? encoder)
+        {
+            return (encoder ?? JavaScriptEncoder.Default).Encode(rawPropertyName);
+        }
+
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/JsonHelpers.Escaping.cs:83]: It doesn't seem like the `encoder` parameter is ever null:
```suggestion
        public static string GetEscapedPropertyName(
            string rawPropertyName,
            JavaScriptEncoder encoder)
        {
            int indexOfFirstCharacterToEncode = JsonWriterHelper.NeedsEscaping(rawPropertyName.AsSpan(), encoder);
            return indexOfFirstCharacterToEncode >= 0
                ? encoder.Encoder(rawPropertyName)
                : rawPropertyName;
        }
```

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs [hunk 8]
```diff
@@ -206,7 +207,7 @@ internal override void GetPath(ref ValueStringBuilder path, JsonNode? child)
                 if (propertyName.AsSpan().ContainsSpecialCharacters())
                 {
                     path.Append("['");
-                    path.Append(propertyName);
+                    path.Append(JsonHelpers.GetPropertyName(propertyName, JavaScriptEncoder.Default));
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:210]:
```suggestion
                    path.Append(JsonHelpers.GetEscapedPropertyName(propertyName, JavaScriptEncoder.Default));
```

**Review by eiriktsarpalis — DISMISSED** [review]
Thanks!

**Comment by eiriktsarpalis** [comment]
> This PR isn't a complete fix.

@gregsdennis Presumably you mean the bracketing of properties starting with the dollar sign? Or is it missing something else?

**Comment by gregsdennis** [comment]
Yes, that.  What's being generated isn't a value path.

If that's not fixed, the issue isn't resolved.

**Comment by eiriktsarpalis** [comment]
> but it does not fix problem with shorthand property.
>I did so because [Newtonsoft.Json 13.0.3](https://github.com/JamesNK/Newtonsoft.Json/releases/tag/13.0.3) have the same behavior:

@Maximys I don't think we need to base our behavior on what Json.NET is doing.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 9]
```diff
@@ -1343,7 +1343,8 @@ private static string FormatJsonSerializerDefaults(JsonSerializerDefaults defaul
             private static string GetCreateValueInfoMethodRef(string typeCompilableName) => $"{CreateValueInfoMethodName}<{typeCompilableName}>";
 
             private static string FormatBool(bool value) => value ? "true" : "false";
-            private static string FormatStringLiteral(string? value) => value is null ? "null" : $"\"{value}\"";
+            private static string FormatStringLiteral(string? value)
+                => value is null ? "null" : SyntaxFactory.Literal(value).ToFullString();
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:1347]:
```suggestion
                => value is null ? "null" : SymbolDisplay.FormatLiteral(value, quote: true);
```

File: src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonNode/ParentPathRootTests.cs [hunk 10]
```diff
@@ -180,5 +192,49 @@ public static void Parent_Object()
             parent.Add("MyProp", child);
             Assert.True(child.Options.Value.PropertyNameCaseInsensitive);
         }
+
+        public static IEnumerable<object[]> GetPath_ShouldReturnExpectedValue_TestData()
+        {
+            yield return new object[]
+            {
+                JsonNode.Parse("""{"$myRoot":{"foo['bar":"baz"}}""")["$myRoot"]["foo['bar"],
+                "$.$myRoot['foo[\\u0027bar']"
```
Comment (gregsdennis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonNode/ParentPathRootTests.cs:201]: If I'm reading these tests right, they're still not creating the correct paths.  `$myRoot` isn't valid for the dot syntax and needs to be bracketed.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.Tests/JsonNode/ParentPathRootTests.cs:201]: @Maximys would be possible to revisit this? Thanks!
